### PR TITLE
Add Support to Generate GraphQL Schema for Local Service Variable Declaration

### PR DESCRIPTION
**Pull request by zetcco**
## Purpose

Fixes: [#3317](https://github.com/ballerina-platform/ballerina-library/issues/3317)

## Examples
This improvement enables to generate GraphQL Schema for the following locally defined `backendService` service object.
```ballerina
import ballerina/lang.runtime;
import ballerina/graphql;

public function main() returns error? {
    // local service variable declaration
    graphql:Service backendService = service object {
        resource function get greeting() returns string {
            return "Hello, World";
        }
    };
    
    graphql:Listener serverEP = check new(4000);
    check serverEP.attach(backendService, "graphql");
    check serverEP.'start();
    runtime:registerListener(serverEP);
}
```

## Checklist
- [x] Linked to an issue
- [x] Updated the changelog
- [x] Added tests
- [ ] Updated the spec
- [ ] Checked native-image compatibility
- [x] No `commons` package changes (if there are any, please update the GraphQL version in [GraphQL tools](https://github.com/ballerina-platform/graphql-tools) and [Ballerina dev tools](https://github.com/ballerina-platform/ballerina-dev-tools))
- [ ] No `compiler` package changes (if there are any, please update the GraphQL version in [Ballerina dev tools](https://github.com/ballerina-platform/ballerina-dev-tools))

### CODE REVIEW

File: compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/LocalLevelServiceObjectAnalysisTask.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) 2022, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (SachinAkash01) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/LocalLevelServiceObjectAnalysisTask.java:2]:
```suggestion
 * Copyright (c) 2024, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```

File: compiler-plugin-tests/src/test/resources/ballerina_sources/generator_tests/25_local_graphql_service_object/service.bal [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    graphql:Service gqlService = service object {
+        resource function get greeting() returns string {
+            return "Wubba lubba dub dub!!";
+        }
+    };
+    graphql:Listener l = check new(9090);
+    check l.attach(gqlService, "graphql");
+    check l.'start();
+    runtime:registerListener(l);
+}
```
Comment (SachinAkash01) [inline compiler-plugin-tests/src/test/resources/ballerina_sources/generator_tests/25_local_graphql_service_object/service.bal:14]:
```suggestion
}

```

File: compiler-plugin-tests/src/test/resources/ballerina_sources/generator_tests/25_local_graphql_service_object/service.bal [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import ballerina/lang.runtime;
```
Comment (SachinAkash01) [inline compiler-plugin-tests/src/test/resources/ballerina_sources/generator_tests/25_local_graphql_service_object/service.bal:1]: Add license header here as well..

Comment (SachinAkash01) [inline compiler-plugin-tests/src/test/resources/ballerina_sources/generator_tests/25_local_graphql_service_object/service.bal:1]: Add the license header on top..

File: compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/LocalLevelServiceObjectAnalysisTask.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        if (serviceValidator.isErrorOccurred()) {
+            return;
+        }
+        Schema schema = generateSchema(context, interfaceEntityFinder, node, null);
+        DocumentId documentId = context.documentId();
+        addToModifierContextMap(documentId, node, schema);
+    }
+
+    public boolean isGraphQLServiceObjectDeclaration(SemanticModel semanticModel,
+                                                            TypedBindingPatternNode typedBindingPatternNode) {
```
Comment (ayeshLK) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/LocalLevelServiceObjectAnalysisTask.java:74]: Shall we properly format this code segment ?

File: compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/LocalLevelServiceObjectAnalysisTask.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        if (moduleSymbol.isEmpty()) {
+            return false;
+        }
+        if (!isPresentAndEquals(moduleSymbol.get().getName(), PACKAGE_NAME)) {
+            return false;
+        }
+        return isPresentAndEquals(typeSymbol.getName(), SERVICE_NAME);
+    }
+
+    private boolean isPresentAndEquals(Optional<String> value, String equals) {
```
Comment (ayeshLK) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/LocalLevelServiceObjectAnalysisTask.java:99]:
```suggestion
    private boolean isPresentAndEquals(Optional<String> actualValue, String expectedValue) {
```

File: compiler-plugin-tests/src/test/resources/ballerina_sources/generator_tests/25_local_graphql_service_object/service.bal [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+import ballerina/lang.runtime;
+import ballerina/graphql;
+
+public function main() returns error? {
+    graphql:Service gqlService = service object {
+        resource function get greeting() returns string {
+            return "Wubba lubba dub dub!!";
+        }
+    };
```
Comment (DimuthuMadushan) [inline compiler-plugin-tests/src/test/resources/ballerina_sources/generator_tests/25_local_graphql_service_object/service.bal:25]: Shall we add a test with complex scenario?

File: compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/GraphqlCodeModifier.java [hunk 7]
```diff
@@ -46,6 +46,8 @@ public void init(CodeModifierContext modifierContext) {
                                                   SyntaxKind.SERVICE_DECLARATION);
         modifierContext.addSyntaxNodeAnalysisTask(
                 new ModuleLevelVariableDeclarationAnalysisTask(this.modifierContextMap), SyntaxKind.MODULE_VAR_DECL);
+        modifierContext.addSyntaxNodeAnalysisTask(
+                new LocalLevelServiceObjectAnalysisTask(this.modifierContextMap), SyntaxKind.OBJECT_CONSTRUCTOR);
```
Comment (ThisaruGuruge) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/GraphqlCodeModifier.java:50]: A better name would be:

```suggestion
                new ObjectConstructorAnalysisTask(this.modifierContextMap), SyntaxKind.OBJECT_CONSTRUCTOR);
```

File: compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/LocalLevelServiceObjectAnalysisTask.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+import io.ballerina.stdlib.graphql.compiler.schema.generator.GraphqlModifierContext;
+import io.ballerina.stdlib.graphql.compiler.service.InterfaceEntityFinder;
+import io.ballerina.stdlib.graphql.compiler.service.validator.ServiceValidator;
+
+import java.util.Map;
+import java.util.Optional;
+
+import static io.ballerina.stdlib.graphql.compiler.Utils.hasCompilationErrors;
+
+public class LocalLevelServiceObjectAnalysisTask extends ServiceAnalysisTask {
```
Comment (ThisaruGuruge) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/LocalLevelServiceObjectAnalysisTask.java:41]:
```suggestion
public class ObjectConstructorAnalysisTask extends ServiceAnalysisTask {
```

File: compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/LocalLevelServiceObjectAnalysisTask.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+        return isGraphqlServiceQualifiedNameReference(semanticModel, (QualifiedNameReferenceNode) typeDescriptorNode);
+    }
+
+    private boolean isGraphqlServiceQualifiedNameReference(SemanticModel semanticModel,
+                                                           QualifiedNameReferenceNode nameReferenceNode) {
+        Optional<Symbol> symbol = semanticModel.symbol(nameReferenceNode);
+        if (symbol.isEmpty()) {
+            return false;
+        }
+        TypeReferenceTypeSymbol typeSymbol = (TypeReferenceTypeSymbol) symbol.get();
```
Comment (ThisaruGuruge) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/LocalLevelServiceObjectAnalysisTask.java:88]:
```suggestion
        if (semanticModel.symbol(nameReferenceNode).isEmpty()) {
            return false;
        }
        TypeReferenceTypeSymbol typeSymbol = (TypeReferenceTypeSymbol) semanticModel.symbol(nameReferenceNode).get();
```

File: compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/LocalLevelServiceObjectAnalysisTask.java [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+        Optional<ModuleSymbol> moduleSymbol = typeSymbol.getModule();
+        if (moduleSymbol.isEmpty()) {
+            return false;
+        }
+        if (!isPresentAndEquals(moduleSymbol.get().getName(), PACKAGE_NAME)) {
+            return false;
+        }
+        return isPresentAndEquals(typeSymbol.getName(), SERVICE_NAME);
```
Comment (ThisaruGuruge) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/LocalLevelServiceObjectAnalysisTask.java:96]:
```suggestion
        if (typeSymbol.getModule().isEmpty()) {
            return false;
        }
        if (!PACKAGE_NAME.equals(typeSymbol.getModule().get().getName()) {
            return false;
        }
        if (typeSymbol.getName().isEmpty()) {
            return false;
        }
        return SERVICE_NAME.equals(type.getName());
```

File: compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/LocalLevelServiceObjectAnalysisTask.java [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    private boolean isPresentAndEquals(Optional<String> actualValue, String expectedValue) {
```
Comment (ThisaruGuruge) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/LocalLevelServiceObjectAnalysisTask.java:99]: A function should do one thing and one thing only. When your function has the word `and`, that's a good hint that the function does more than one thing.

File: compiler-plugin-tests/src/test/resources/ballerina_sources/generator_tests/25_local_graphql_service_object/service.bal (hunk 6 above)
Comment (zetcco) [reply compiler-plugin-tests/src/test/resources/ballerina_sources/generator_tests/25_local_graphql_service_object/service.bal:25]: Maybe something like this?
```ballerina
type Foo record {|
    string name;
    int age;
|};

public function main() returns error? {
    graphql:Service gqlService = service object {
        resource function get number() returns int {
            return 10;
        }
        
        resource function get greeting() returns string {
            return "Hello";
        }
        
        resource function get foo() returns Foo {
            return {
                name: "Bar",
                age: 100
            };
        }
    };
    graphql:Listener l = check new(9090);
    check l.attach(gqlService, "graphql");
    check l.'start();
    runtime:registerListener(l);
}
```

File: compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/schema/generator/GraphqlSourceModifier.java [hunk 12]
```diff
@@ -146,6 +148,23 @@ private ModulePartNode modifyDocument(SourceModifierContext context, ModulePartN
...
+                            = (ObjectConstructorExpressionNode) targetNode;
+                    ObjectConstructorExpressionNode updatedGraphqlServiceObject = modifyServiceObjectNode(
+                            graphqlServiceVariableDeclaration, schemaString, prefix);
+                    VariableDeclarationNode parentNode
+                            = (VariableDeclarationNode) graphqlServiceVariableDeclaration.parent();
+                    VariableDeclarationNode updatedParentNode = parentNode.modify()
+                                                                          .withInitializer(updatedGraphqlServiceObject)
+                                                                          .apply();
+                    Node replacingNode = updatedParentNode;
+                    NonTerminalNode iNode = parentNode;
```
Comment (ThisaruGuruge) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/schema/generator/GraphqlSourceModifier.java:162]: `iNode` does not explain the purpose of the variable correctly. Shall we use a better name?

File: compiler-plugin-tests/src/test/resources/ballerina_sources/generator_tests/25_local_graphql_service_object/service.bal (hunk 6 above)
Comment (ThisaruGuruge) [reply compiler-plugin-tests/src/test/resources/ballerina_sources/generator_tests/25_local_graphql_service_object/service.bal:25]: +1. We need a test where we have a service declared with global variable and a local variable. We also need a test with a service declared using listener binding, and a locally defined service. We need to test whether the schema is generated correctly for all the services in the above scenarios. The point is, when we change the syntax tree, we need to make sure we do not remove the changes from the other services.

File: ballerina-tests/Dependencies.toml
Comment (ThisaruGuruge) [inline ballerina-tests/Dependencies.toml:1]: Let's add some runtime tests with local variable service declaration.

File: compiler-plugin-tests/src/test/resources/ballerina_sources/generator_tests/25_local_service_object_declarations/service.bal [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+    };
+    graphql:Listener localListener1 = check new(9090);
+    graphql:Listener localListener2 = check new(9091);
+    check localListener1.attach(localService1);
+    check localListener2.attach(localService2);
+    check localListener1.'start();
+    check localListener2.'start();
+    runtime:registerListener(localListener1);
+    runtime:registerListener(localListener2);
+}
```
Comment (DimuthuMadushan) [inline compiler-plugin-tests/src/test/resources/ballerina_sources/generator_tests/25_local_service_object_declarations/service.bal:39]:
```suggestion
}

```

Comment (DimuthuMadushan) [reply compiler-plugin-tests/src/test/resources/ballerina_sources/generator_tests/25_local_service_object_declarations/service.bal:39]: check other places as well.

File: compiler-plugin-tests/src/test/resources/ballerina_sources/generator_tests/26_object_field_service_object_declarations/service.bal [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+        check localListener2.'start();
+        runtime:registerListener(localListener1);
+        runtime:registerListener(localListener2);
+    }
+}
+
+public function main() returns error? {
+    TestService serviceClass = new ();
+    check serviceClass.startService();
+}
```
Comment (DimuthuMadushan) [inline compiler-plugin-tests/src/test/resources/ballerina_sources/generator_tests/26_object_field_service_object_declarations/service.bal:49]: ditto

File: compiler-plugin-tests/src/test/resources/ballerina_sources/generator_tests/27_inline_service_declarations/service.bal [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/lang.runtime; import ballerina/graphql; import ballerina/http; service / on new http:Listener(9090) { resource function get greet() returns string { return "Hello from global http service"; } } public function main() returns error? { graphql:Service gqlService = service object { resource function get greet() returns string { return "Hello from gql service"; } }; http:Service httpService = service object { resource function get greet() returns string { return "Hello from local http service"; } }; graphql:Listener gqlListener = check new(9091); http:Listener httpListener = check new(9092); check gqlListener.attach(gqlService); check httpListener.attach(httpService); check gqlListener.'start(); check httpListener.'start(); runtime:registerListener(gqlListener); runtime:registerListener(httpListener); }
```
Comment (MohamedSabthar) [inline compiler-plugin-tests/src/test/resources/ballerina_sources/generator_tests/27_inline_service_declarations/service.bal:17]: shall we format this file?

Comment (zetcco) [reply compiler-plugin-tests/src/test/resources/ballerina_sources/generator_tests/27_inline_service_declarations/service.bal:17]: This was intentional, as the code modifier now utilizes the text range to modify the source code.

Comment (MohamedSabthar) [reply compiler-plugin-tests/src/test/resources/ballerina_sources/generator_tests/27_inline_service_declarations/service.bal:17]: On the other hand, lets keep this as we are modifying the text range in the code

**Review by ThisaruGuruge — COMMENTED** [review]
Shall we add test in the `ballerina-tests` module, so that we can test the runtime as well?

**Review by ThisaruGuruge — APPROVED** [review]

**Pull request merged**